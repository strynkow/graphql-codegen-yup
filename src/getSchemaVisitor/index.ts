import { DeclarationBlock, indent } from '@graphql-codegen/visitor-plugin-common';
import {
  EnumTypeDefinitionNode,
  GraphQLSchema,
  InputObjectTypeDefinitionNode,
  InterfaceTypeDefinitionNode,
  ObjectTypeDefinitionNode,
} from 'graphql';

import { TsVisitor } from '@graphql-codegen/typescript';
import { ValidationSchemaPluginConfig } from '../types';
import getFieldSchema from './getFieldSchema';

export default function getSchemaVisitor(schema: GraphQLSchema, config: ValidationSchemaPluginConfig) {
  const tsVisitor = new TsVisitor(schema, config);

  const importTypes: string[] = [];

  return {
    InputObjectTypeDefinition: (node: InputObjectTypeDefinitionNode) => {
      const name = tsVisitor.convertName(node.name.value);
      importTypes.push(name);

      const shape = node.fields
        ?.map((field) => getFieldSchema({ config, tsVisitor, schema, field, indentCount: 2 }))
        .join(',\n');

      return new DeclarationBlock({})
        .export()
        .asKind('function')
        .withName(`${name}Schema(): yup.SchemaOf<${name}>`)
        .withBlock([indent(`return yup.object({`), shape, indent('})')].join('\n')).string;
    },
    ObjectTypeDefinition: (node: ObjectTypeDefinitionNode) => {
      const name = tsVisitor.convertName(node.name.value);
      importTypes.push(name);

      const shape = node.fields
        ?.map((field) => getFieldSchema({ config, tsVisitor, schema, field, indentCount: 2 }))
        .join(',\n');

      return new DeclarationBlock({})
        .export()
        .asKind('function')
        .withName(`${name}Schema(): yup.SchemaOf<${name}>`)
        .withBlock(
          [
            indent(`return yup.object({`),
            indent(`__typename: yup.mixed().oneOf(['${node.name.value}', undefined]),`, 2),
            shape,
            indent('})'),
          ].join('\n')
        ).string;
    },
    EnumTypeDefinition: (node: EnumTypeDefinitionNode) => {
      const enumname = tsVisitor.convertName(node.name.value);
      importTypes.push(enumname);

      if (config.enumsAsTypes) {
        return new DeclarationBlock({})
          .export()
          .asKind('const')
          .withName(`${enumname}Schema`)
          .withContent(
            `yup.mixed().oneOf([${node.values?.map((enumOption) => `'${enumOption.name.value}'`).join(', ')}])`
          ).string;
      }

      const values = node.values
        ?.map(
          (enumOption) =>
            `${enumname}.${tsVisitor.convertName(enumOption.name, {
              useTypesPrefix: false,
              transformUnderscore: true,
            })}`
        )
        .join(', ');
      return new DeclarationBlock({})
        .export()
        .asKind('const')
        .withName(`${enumname}Schema`)
        .withContent(`yup.mixed().oneOf([${values}])`).string;
    },
    InterfaceTypeDefinition: (node: InterfaceTypeDefinitionNode) => {
      const name = tsVisitor.convertName(node.name.value);
      importTypes.push(name);

      const shape = node.fields
        ?.map((field) => getFieldSchema({ config, tsVisitor, schema, field, indentCount: 2 }))
        .join(',\n');

      return new DeclarationBlock({})
        .export()
        .asKind('function')
        .withName(`${name}Schema(): yup.SchemaOf<${name}>`)
        .withBlock(
          [
            indent(`return yup.object({`),
            indent(`__typename: yup.mixed().oneOf(['${node.name.value}', undefined]),`, 2),
            shape,
            indent('})'),
          ].join('\n')
        ).string;
    },
  };
}
