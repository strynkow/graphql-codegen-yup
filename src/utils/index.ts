import { ListTypeNode, NamedTypeNode, NonNullTypeNode, TypeNode } from 'graphql';

export function isListType(typ?: TypeNode): typ is ListTypeNode {
  return typ?.kind === 'ListType';
}

export function isNonNullType(typ?: TypeNode): typ is NonNullTypeNode {
  return typ?.kind === 'NonNullType';
}

export function isNamedType(typ?: TypeNode): typ is NamedTypeNode {
  return typ?.kind === 'NamedType';
}

export function isInput(kind: string): boolean {
  return kind.includes('Input');
}

export function maybeLazy(type: TypeNode, schema: string): string {
  if (isNamedType(type) && isInput(type.name.value)) {
    // TODO (orange): We have the tech to change this, after getting problems solved.
    return `yup.lazy(() => ${schema}) as never`;
  }
  return schema;
}
