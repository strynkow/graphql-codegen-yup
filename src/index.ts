import { PluginFunction, Types, oldVisit } from '@graphql-codegen/plugin-helpers';

import { GraphQLSchema } from 'graphql';
import { ValidationSchemaPluginConfig } from './types';
import getSchemaVisitor from './getSchemaVisitor';
import { transformSchemaAST } from '@graphql-codegen/schema-ast';

export const plugin: PluginFunction<ValidationSchemaPluginConfig, Types.ComplexPluginOutput> = (
  schema: GraphQLSchema,
  _documents: Types.DocumentFile[],
  config: ValidationSchemaPluginConfig
): Types.ComplexPluginOutput => {
  const { schema: _schema, ast } = transformSchemaAST(schema, config);
  const visitor = getSchemaVisitor(_schema, config);

  const result = oldVisit(ast, {
    leave: visitor,
  });

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  const generated = result.definitions.filter((def) => typeof def === 'string');

  return {
    prepend: [`import * as yup from 'yup';`],
    content: generated.join('\n'),
  };
};
