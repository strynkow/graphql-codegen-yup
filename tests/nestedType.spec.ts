import { buildSchema } from 'graphql';
import { plugin } from '../src/index';

describe('Nested types', () => {
  it('compiles inputs', async () => {
    const schema = buildSchema(/* GraphQL */ `
      input Something {
        str: String!
      }
      input OtherThing {
        something: Something!
      }
    `);
    const result = await plugin(schema, [], {}, {});
    expect(result.content).toContain('str: yup.string().defined()');
    expect(result.content).toContain('something: SomethingSchema().defined()');
    expect(result.content).toMatchSnapshot();
  });

  it('compiles types', async () => {
    const schema = buildSchema(/* GraphQL */ `
      type Whatever {
        str: String!
        num: Float!
      }

      type RunningOutOfNames {
        whateverParam: Whatever!
        num: Float!
      }
    `);
    const result = await plugin(schema, [], {}, {});
    expect(result.content).toContain('whateverParam: WhateverSchema().defined()');
    expect(result.content).toContain('num: yup.number().defined()');
    expect(result.content).toMatchSnapshot();
  });
});
