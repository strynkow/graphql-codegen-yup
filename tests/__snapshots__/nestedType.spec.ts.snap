// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nested types compiles inputs 1`] = `
"export function SomethingSchema(): yup.SchemaOf<Something> {
  return yup.object({
    str: yup.string().defined()
  })
}

export function OtherThingSchema(): yup.SchemaOf<OtherThing> {
  return yup.object({
    something: SomethingSchema().defined()
  })
}
"
`;

exports[`Nested types compiles types 1`] = `
"export function WhateverSchema(): yup.SchemaOf<Whatever> {
  return yup.object({
    __typename: yup.mixed().oneOf(['Whatever', undefined]),
    str: yup.string().defined(),
    num: yup.number().defined()
  })
}

export function RunningOutOfNamesSchema(): yup.SchemaOf<RunningOutOfNames> {
  return yup.object({
    __typename: yup.mixed().oneOf(['RunningOutOfNames', undefined]),
    whateverParam: WhateverSchema().defined(),
    num: yup.number().defined()
  })
}
"
`;
